demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,675 @@
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "ERROR: %d - Failed to Verify ECDSA.\r\n", xResult ) );
+            }
+
+            mbedtls_mpi_free( &xR );
+            mbedtls_mpi_free( &xS );
+
+            /*//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////// */
+
+
+            /* Generate a PK Context for CSR */
+            /* Getting header info from ECKEY type */
+            mbedtls_pk_type_t type_key = MBEDTLS_PK_ECKEY;
+            const mbedtls_pk_info_t * header = mbedtls_pk_info_from_type( type_key );
+
+            /* Creating copy of header to pass in custom sign function (original header is immutable) */
+            mbedtls_pk_info_t * header_copy = pvPortMalloc( sizeof( mbedtls_pk_info_t ) );
+
+            if( header_copy == NULL )
+            {
+                configPRINTF( ( "ERROR: Failed to allocate memory for header_copy variable. CSR cannot be generated\r\n" ) );
+                return;
+            }
+
+            memcpy( header_copy, header, sizeof( struct mbedtls_pk_info_t ) );
+
+            header_copy->sign_func = &prvPrivateKeySigningCallback;
+
+            /* Initializing PK Context */
+            mbedtls_pk_context pk_cont;
+            mbedtls_pk_init( &pk_cont );
+            ret = mbedtls_pk_setup( &pk_cont, header_copy );
+
+            if( ret != 0 )
+            {
+                configPRINTF( ( "ERROR: %d - Failed to initialize PK context with given information.\r\n", ret ) );
+            }
+
+            pk_cont.pk_ctx = &xEcdsaContext;
+
+
+            /* Generating CSR */
+            /* Initializing CSR Context */
+            mbedtls_x509write_csr my_csr;
+            mbedtls_x509write_csr_init( &my_csr );
+            ret = mbedtls_x509write_csr_set_subject_name( &my_csr, "CN=ThingName" ); /* This name is configurable to your personal thing name. */
+
+            if( ret != 0 )
+            {
+                configPRINTF( ( "ERROR: %d - Failed to set subject name of CSR context.\r\n", ret ) );
+            }
+
+            mbedtls_x509write_csr_set_key( &my_csr, &pk_cont );
+            mbedtls_x509write_csr_set_md_alg( &my_csr, MBEDTLS_MD_SHA256 );
+            ret = mbedtls_x509write_csr_set_key_usage( &my_csr, MBEDTLS_X509_KU_DIGITAL_SIGNATURE );
+
+            if( ret != 0 )
+            {
+                configPRINTF( ( "ERROR: %d - Failed to set key usage of CSR context.\r\n", ret ) );
+            }
+
+            ret = mbedtls_x509write_csr_set_ns_cert_type( &my_csr, MBEDTLS_X509_NS_CERT_TYPE_SSL_CLIENT );
+
+            if( ret != 0 )
+            {
+                configPRINTF( ( "ERROR: %d - Failed to set NS Cert Type of CSR context.\r\n", ret ) );
+            }
+
+            /* Output Buffer */
+            unsigned char * final_csr = pvPortMalloc( 2000 );
```

Comment:
A nicer way to handle this would be to have an inline #define for the length of the buffer. Then declare len_buf and initialize it to the value of the #define. Then call malloc, passing in len_buf. In my experience, that ordering is less likely to lead to mistakes down the road.

Reply:
Thank you. I added the #define for the buffer size and used that instead.